kernel/src/fs/epoll/epoll_file.rs

```diff
@@ -51,113 +51,120 @@ impl EpollFile {
-
         // Add the new entry to the interest list and start monitoring its events
-        let mut interest = self.interest.lock();
-        if interest.contains_key(&fd) {
-            return_errno_with_message!(Errno::EEXIST, "the fd has been added");
-        }
-        file.register_observer(entry.self_weak() as _, IoEvents::all())?;
-        interest.insert(fd, entry.clone());
-        // Register self to the file table entry
-        file_table_entry.register_observer(self.weak_self.clone() as _);
-        let file = file.clone();
-        drop(file_table);
-        drop(interest);
+        let entry = {
+            let mut interest = self.interest.lock();
+
+            if interest.contains(&(fd, KeyableWeak::from(Arc::downgrade(&file)))) {
+                return_errno_with_message!(
+                    Errno::EEXIST,
+                    "the file is already in the interest list"
+                );
+            }
+
+            let entry = EpollEntry::new(fd, &file, ep_event, ep_flags, self.weak_self.clone())?;
+            let inserted = interest.insert(entry.clone().into());
+            assert!(inserted);
+
+            entry
+        };
 
         // Add the new entry to the ready list if the file is ready
-        let events = file.poll(mask, None);
+        let events = file.poll(ep_event.events, None);
         if !events.is_empty() {
             self.push_ready(entry);
         }
+
         Ok(())
     }
 
-    fn del_interest(&self, fd: FileDesc) -> Result<()> {
+    fn del_interest(&self, fd: FileDesc, file: KeyableWeak<dyn FileLike>) -> Result<()> {
         let mut interest = self.interest.lock();
-        let entry = interest
-            .remove(&fd)
-            .ok_or_else(|| Error::with_message(Errno::ENOENT, "fd is not in the interest list"))?;
+
+        let KeyableEntry(entry) = interest.take(&(fd, file)).ok_or_else(|| {
+            Error::with_message(Errno::ENOENT, "the file is not in the interest list")
+        })?;
 
         // If this epoll entry is in the ready list, then we should delete it.
         // But unfortunately, deleting an entry from the ready list has a
         // complexity of O(N).
         //
-        // To optimize the performance, we only mark the epoll entry as
-        // deleted at this moment. The real deletion happens when the ready list
-        // is scanned in EpolFile::wait.
-        entry.set_deleted();
-
-        let file = match entry.file() {
-            Some(file) => file,
-            // TODO: should we warn about it?
-            None => return Ok(()),
-        };
+        // To optimize performance, we postpone the actual deletion to the time
+        // when the ready list is scanned in `EpolFile::wait`. This can be done
+        // because the strong reference count will reach zero and `Weak::upgrade`
+        // will fail.
+
+        drop(interest);
+        drop(entry);
```

Comment:
Is there any special reason to add two `drop` here? They can also be added by compiler, right?

Reply:
I previously want to give a hint that something is going to happen in the `Drop` impl. This is because I move the `unregister_observer()` call to the `Drop` impl of `EpollEntry`, a change that seems to be straightforward.

However, when I revisit the code, I realize there may be some problems. In the event observer callback, we need to hold the lock of the ready list. At this point, we also hold the lock on the pollee (`Pollee`/`Subject`), which happens before the callback is invoked.

This means that as long as we are holding the ready list lock, we cannot try to do `unregister_observer`, otherwise we will be trying to hold the two locks (the ready list lock and the pollee lock) in reverse order. This has the potential to cause deadlocks.

Here are some facts about this problem:
 - The problem existed prior to this PR. We're trying to call `del_interest` directly in `pop_ready` to handle `EPOLLONESHOT`.
   - However, this is just a completely wrong implementation, since `EPOLLONESHOT` means that the interest is _disabled_ after being armed, not deleted.
 - Moving `unregister_observer` to the `Drop` impl of `EpollFile` makes the problem more complicated and harder to notice. Note that although the ready list stores `Weak<EpollFile>`, they are upgraded to `Arc<EpollFile>` in `pop_ready`, and they can be the last references, so they can be dropped while holding the ready list lock.

I think the newly added commit, https://github.com/asterinas/asterinas/pull/1277/commits/d980e84064302009c60a0e1b743c93c9ac93dfd2, should resolve these problems. Could you please have a look at it?

Reply:
A slightly revised way to solve this problem is to:
 1. Renaming `KeyableEntry` to `EpollEntryHolder` (in response to the name confusion mentioned in the comment https://github.com/asterinas/asterinas/pull/1277#discussion_r1741355500).
 2. Making a `Drop` impl for `EpollEntryHolder`, which calls `unregister_observer` (this drop can only happen in `del_interest`).
 
This should also work, see https://github.com/asterinas/asterinas/compare/d980e84064302009c60a0e1b743c93c9ac93dfd2..3d9405460fe7dd25f34d08487e8f9c9e8f066b31.

Reply:
> Making a Drop impl for EpollEntryHolder, which calls unregister_observer (this drop can only happen in del_interest).

This is a clever solution to avoid unregister observer when dropping `EpollFile` from upgrade! 

> Renaming KeyableEntry to EpollEntryHolder

The new name looks fine to me!